src/Utilities/Compiler/Options/SymbolNamesOption.cs

```diff
@@ -98,32 +154,145 @@ private SymbolNamesOption()
                 }
             }
 
-            if (namesBuilder.Count == 0 && symbolsBuilder.Count == 0)
+            if (namesBuilder.Count == 0 && symbolsBuilder.Count == 0 && wildcardNamesBuilder.Count == 0)
             {
                 return Empty;
             }
 
-            return new SymbolNamesOption(namesBuilder.ToImmutableDictionaryAndFree(), symbolsBuilder.ToImmutableDictionaryAndFree());
+            return new SymbolNamesOption(namesBuilder.ToImmutableDictionaryAndFree(), symbolsBuilder.ToImmutableDictionaryAndFree(),
+                wildcardNamesBuilder.ToImmutableDictionaryAndFree(x => x.Key, x => x.Value.ToImmutableDictionaryAndFree(), wildcardNamesBuilder.Comparer));
         }
 
         public bool IsEmpty => ReferenceEquals(this, Empty);
 
         public bool Contains(ISymbol symbol)
-            => _symbols.ContainsKey(symbol) || _names.ContainsKey(symbol.Name);
+            => _symbols.ContainsKey(symbol) || _names.ContainsKey(symbol.Name) || TryGetFirstWildcardMatch(symbol, out _);
 
         /// <summary>
         /// Gets the value associated with the specified symbol in the option specification.
         /// </summary>
-        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out string? value) =>
-            _symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value);
+        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out string? value)
+        {
+            if (_symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value))
+            {
+                return true;
+            }
+
+            if (TryGetFirstWildcardMatch(symbol, out var match))
+            {
+                value = match.Value;
+                return true;
+            }
+
+            value = null;
+            return false;
+        }
 
         public override bool Equals(object obj) => Equals(obj as SymbolNamesOption);
 
         public bool Equals(SymbolNamesOption? other)
-            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols);
+            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols) && _wildcardNamesBySymbolKind.IsEqualTo(other._wildcardNamesBySymbolKind);
 
         public override int GetHashCode()
-            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols));
+            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols), HashUtilities.Combine(_wildcardNamesBySymbolKind));
+
+        private bool TryGetFirstWildcardMatch(ISymbol symbol, out KeyValuePair<string, string?> firstMatch)
```

Comment:
This is the "simplest" algo I have found to match the various cases I wanted to handle. I don't think this performs particularly well. If we decided to stick with this algo we might want to have some cache of symbol <-> result to avoid doing the full match everytime.

Reply:
Yeah, agreed, we need to build up a cache to avoid iterating over the same containing types and namespaces again. I think you can use data structures and logic similar to below used in RestrictedIVT analyzer: https://github.com/dotnet/roslyn-analyzers/blob/3cc0345042a39d23c8b82fd02d0d2dd298dc92cb/src/Microsoft.CodeAnalysis.BannedApiAnalyzers/Core/RestrictedInternalsVisibleToAnalyzer.cs#L196-L258

Maintain a lazyily computed dictionary, say `ConcurrentDictionary<INamespaceOrTypeSymbol, /*isWildCardMatch*/bool>`, and update it with true or false value whenever you encounter a new containing type or namespace.

Reply:
I think rest of the logic looks fine. We can perf tune the rest if we see performance issues.